util/ranger/detacher.go

```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
 	}
 	return mergedDNFItems
 }
+
+// AddGcColumnCond add the `tidb_shard(x) = xxx` to the condition
+// @param[in] cols          the columns of shard index, such as [tidb_shard(a), a, ...]
+// @param[in] accessCond    the conditions relative to the index and arranged by the index column order.
+//                          e.g. the index is uk(tidb_shard(a), a, b) and the where clause is
+//                          `WHERE b = 1 AND a = 2 AND c = 3`, the param accessCond is {a = 2, b = 1} that is
+//                          only relative to uk's columns.
+// @param[in] columnValues  the values of index columns in param accessCond. if accessCond is {a = 2, b = 1},
+//                          columnValues is {2, 1}. if accessCond the "IN" function like `a IN (1, 2)`, columnValues
+//                          is empty.
+// @retval -  []expression.Expression   the new condtions after adding `tidb_shard() = xxx` prefix
+//            error                     if error gernerated, return error
+func AddGcColumnCond(sctx sessionctx.Context,
+	cols []*expression.Column,
+	accessesCond []expression.Expression,
+	columnValues []*valueInfo) ([]expression.Expression, error) {
+
+	if cond := accessesCond[1]; cond != nil {
+		if f, ok := cond.(*expression.ScalarFunction); ok {
+			switch f.FuncName.L {
+			case ast.EQ:
+				return AddGcColumn4EqCond(sctx, cols, accessesCond, columnValues)
+			case ast.In:
+				return AddGcColumn4InCond(sctx, cols, accessesCond)
+			}
+		}
+	}
+
+	return accessesCond, nil
+}
+
+// AddGcColumn4InCond add the `tidb_shard(x) = xxx` for `IN` condition
+// For param explanation, please refer to the function `AddGcColumnCond`.
+// @retval -  []expression.Expression   the new condtions after adding `tidb_shard() = xxx` prefix
+//            error                     if error gernerated, return error
+func AddGcColumn4InCond(sctx sessionctx.Context,
+	cols []*expression.Column,
+	accessesCond []expression.Expression) ([]expression.Expression, error) {
+
+	var errRes error
+	var newAccessCond []expression.Expression
+	record := make([]types.Datum, 1)
+
+	expr := cols[0].VirtualExpr.Clone()
+	andType := types.NewFieldType(mysql.TypeTiny)
+
+	sf := accessesCond[1].(*expression.ScalarFunction)
+	c := sf.GetArgs()[0].(*expression.Column)
+	var AndOrExpr expression.Expression
+	for i, arg := range sf.GetArgs()[1:] {
+		// get every const value and calculate tidb_shard(val)
+		con := arg.(*expression.Constant)
+		conVal, err := con.Eval(chunk.Row{})
+		if err != nil {
+			return accessesCond, err
+		}
+
+		record[0] = conVal
+		mutRow := chunk.MutRowFromDatums(record)
+		exprVal, err := expr.Eval(mutRow.ToRow())
+		if err != nil {
+			return accessesCond, err
+		}
+
+		// tmpArg1 is like `tidb_shard(a) = 8`, tmpArg1 is like `a = 100`
+		exprCon := &expression.Constant{Value: exprVal, RetType: cols[0].RetType}
+		tmpArg1, err1 := expression.NewFunction(sctx, ast.EQ, cols[0].RetType, cols[0], exprCon)
```

Comment:
The err can be directly reused.
We don't need to `err1`, `err2`, and `err3`. Just one `err` is okay.